### PR TITLE
Fix bug of link query order on markdown render

### LINKED ISSUE #14090: urls in markdown are broken - arguments are sorted
## Problem
When I use this markdown 

```md
Bug markdown links
 - https://example.com/?route=action&id=999
 - <https://example.com/?route=action&id=999>
 - [https://example.com/?route=action&id=999](https://example.com/?route=action&id=999)
```

I get in html

```
<p>Bug markdown links</p>
<ul>
<li><a href="https://example.com/?id=999&amp;route=action" rel="nofollow">https://example.com/?route=action&amp;id=999</a></li>
<li><a href="https://example.com/?id=999&amp;route=action" rel="nofollow">https://example.com/?route=action&amp;id=999</a></li>
<li><a href="https://example.com/?id=999&amp;route=action" rel="nofollow">https://example.com/?route=action&amp;id=999</a></li>
</ul>
```

As you can see `href` is malformed (order of arguments is wrong).

## try.gitea.io

https://try.gitea.io/hrv3/issue-14090/src/branch/master/bug.md

## Versions
Gitea: 1.13.0

### PR DESCRIPTION
**Pull request by lunny**
Should fix #14090

We temporarily use a replace on go.mod before https://github.com/microcosm-cc/bluemonday/pull/105 merged.

### CODE REVIEW

**Review by noerw — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@lunny need to change tests:
```
--- FAIL: TestRender_links (0.02s)
    html_test.go:89: 
        	Error Trace:	html_test.go:89
        	            				html_test.go:125
        	Error:      	Not equal: 
        	            	expected: "<p><a href=\"https://github.com/go-gitea/gitea/?p=aaa%2Fbbb.html#ccc-ddd\" rel=\"nofollow\">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>"
        	            	actual  : "<p><a href=\"https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd\" rel=\"nofollow\">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-<p><a href="https://github.com/go-gitea/gitea/?p=aaa%2Fbbb.html#ccc-ddd" rel="nofollow">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>
        	            	+<p><a href="https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd" rel="nofollow">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>
        	Test:       	TestRender_links
    html_test.go:89: 
        	Error Trace:	html_test.go:89
        	            				html_test.go:143
        	Error:      	Not equal: 
        	            	expected: "<p><a href=\"magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	actual  : "<p><a href=\"magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-<p><a href="magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	            	+<p><a href="magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	Test:       	TestRender_links
FAIL
```

**Comment by zeripath** [comment]
no - don't change the tests - fix the query parameter it's not escaping properly.

**Comment by lunny** [comment]
The first test should be right and I changed the bluemonday codes.

I also changed the second test on this PR because it should be a wrong test.

```go
html_test.go:143
        	Error:      	Not equal: 
        	            	expected: "<p><a href=\"magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	actual  : "<p><a href=\"magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-<p><a href="magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	            	+<p><a href="magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	Test:       	TestRender_links
```

It's obvious, the query order changed but should not.

File: vendor/github.com/microcosm-cc/bluemonday/sanitize.go [hunk 1]
```diff
@@ -122,22 +122,79 @@ func escapeUrlComponent(val string) string {
...
-	sanitizedQueryValues := make(url.Values, 0)
-	queryValues := u.Query()
-	for k, vals := range queryValues {
-		sk := html.EscapeString(k)
-		for _, v := range vals {
-			sv := v
-			sanitizedQueryValues.Add(sk, sv)
-		}
+	for i, query := range queryValues {
+		queryValues[i].Key = html.EscapeString(query.Key)
```
Comment (zeripath) [inline vendor/github.com/microcosm-cc/bluemonday/sanitize.go:195]: I don't understand this. What's the aim of this? url.QueryEscape should do for both the key and value.

html.EscapeString will still need to be run on the whole url.

Comment (lunny) [reply vendor/github.com/microcosm-cc/bluemonday/sanitize.go:195]: Hm, I haven't look into this, but just translate from the old code of bluemonday. See https://github.com/microcosm-cc/bluemonday/pull/105/files

**Review by zeripath — APPROVED** [review]

**Pull request merged**